pkg/core/resources/apis/meshservice/generate/generator.go

```diff
@@ -0,0 +1,238 @@
+		if !hasProtocol {
+			appProtocol = core_mesh.ProtocolTCP
+		}
+		port := meshservice_api.Port{
+			Port:        inbound.Port,
+			TargetPort:  intstr.FromInt(int(inbound.Port)),
+			AppProtocol: core_mesh.Protocol(appProtocol),
+		}
+		portsByService[serviceTagValue] = append(portsByService[serviceTagValue], port)
+	}
+
+	services := map[string]*meshservice_api.MeshService{}
+	for serviceTag, ports := range portsByService {
+		ms := meshservice_api.MeshService{
+			Selector: meshservice_api.Selector{
+				DataplaneTags: meshservice_api.DataplaneTags{
+					mesh_proto.ServiceTag: serviceTag,
+				},
+			},
+			Ports: ports,
+		}
+		services[serviceTag] = &ms
+	}
+	return services
+}
+
+type dataplaneAndMeshService struct {
+	dataplane   model.ResourceMeta
+	meshService *meshservice_api.MeshService
+}
+
+// checkMeshServicesConsistency returns a list of dataplanes that conflict and
+// the chosen meshService
+func checkMeshServicesConsistency(
+	meshService *meshservice_api.MeshService,
+	generated []dataplaneAndMeshService,
+) ([]dataplaneAndMeshService, *meshservice_api.MeshService) {
+	if len(generated) == 0 {
+		return nil, nil
+	}
+	if meshService == nil {
+		meshService = generated[0].meshService
+	}
+	var conflicting []dataplaneAndMeshService
+	for _, generatedMeshService := range generated {
+		if servicesDiffer(meshService, generatedMeshService.meshService) {
+			conflicting = append(conflicting, generatedMeshService)
+		}
+	}
+	if len(generated) == len(conflicting) {
+		// None of the generated MeshServices match the existing, drop it and
+		// try again with the first generated. We know we won't recurse
+		// indefinitely because generated[0] won't conflict
+		return checkMeshServicesConsistency(generated[0].meshService, generated)
+	}
+	return conflicting, meshService
+}
+
+func sortDataplanes(dps []*core_mesh.DataplaneResource) []*core_mesh.DataplaneResource {
+	sorted := slices.Clone(dps)
+	slices.SortFunc(sorted, func(a, b *core_mesh.DataplaneResource) int {
+		if a, b := a.Meta.GetCreationTime(), b.Meta.GetCreationTime(); a.Before(b) {
+			return -1
+		} else if a.After(b) {
+			return 1
+		}
+		return strings.Compare(a.Meta.GetName(), b.Meta.GetName())
+	})
+	return sorted
+}
+
+func servicesDiffer(a, b *meshservice_api.MeshService) bool {
+	return !reflect.DeepEqual(a.Ports, b.Ports)
+}
+
+func (g *Generator) generate(ctx context.Context, mesh string, dataplanes []*core_mesh.DataplaneResource, meshServices []*meshservice_api.MeshServiceResource) {
+	log := g.logger.WithValues("mesh", mesh)
+	meshservicesByName := map[string][]dataplaneAndMeshService{}
+	for _, dataplane := range sortDataplanes(dataplanes) {
+		for name, ms := range g.meshServicesForDataplane(dataplane) {
+			meshservicesByName[name] = append(meshservicesByName[name], dataplaneAndMeshService{
+				dataplane:   dataplane.GetMeta(),
+				meshService: ms,
+			})
+		}
+	}
+
+	for _, meshService := range meshServices {
```

Comment:
should we only consider services with label `managed-by: meshservice-generator`?

Reply:
done